xbmc/utils/SeekHandler.cpp

```diff
@@ -298,9 +307,123 @@ bool CSeekHandler::OnAction(const CAction &action)
         Seek(action.GetID() == ACTION_ANALOG_SEEK_FORWARD, action.GetAmount(), action.GetRepeat(), true);
       return true;
     }
+    case REMOTE_0:
+    case REMOTE_1:
+    case REMOTE_2:
+    case REMOTE_3:
+    case REMOTE_4:
+    case REMOTE_5:
+    case REMOTE_6:
+    case REMOTE_7:
+    case REMOTE_8:
+    case REMOTE_9:
+    {
+      if (!g_application.CurrentFileItem().IsLiveTV())
+      {
+        ChangeTimeCode(action.GetID());
+        return true;
+      }
+    }
+    break;
     default:
       break;
   }
 
   return false;
 }
+
+bool CSeekHandler::SeekTimeCode(const CAction &action)
+{
+  if (m_timeCodePosition <= 0)
+    return false;
+
+  switch (action.GetID())
+  {
+    case ACTION_SELECT_ITEM:
+    case ACTION_PLAYER_PLAY:
+    case ACTION_PAUSE:
+    {
+      CSingleLock lock(m_critSection);
+
+      g_application.m_pPlayer->SeekTime(GetTimeCodeAmount() * 1000);
+      Reset();
+      return true;
+    }
+    case ACTION_SMALL_STEP_BACK:
+    case ACTION_STEP_BACK:
+    case ACTION_BIG_STEP_BACK:
+    case ACTION_CHAPTER_OR_BIG_STEP_BACK:
+    {
+      SeekSeconds(-GetTimeCodeAmount());
+      return true;
+    }
+    case ACTION_STEP_FORWARD:
+    case ACTION_BIG_STEP_FORWARD:
+    case ACTION_CHAPTER_OR_BIG_STEP_FORWARD:
+    {
+      SeekSeconds(GetTimeCodeAmount());
+      return true;
+    }
+    default:
+      break;
+  }
+  return false;
+}
+
+void CSeekHandler::ChangeTimeCode(int remote)
+{
+  if (remote >= REMOTE_0 && remote <= REMOTE_9)
+  {
+    m_timerTimeCode.StartZero();
+
+    if (m_timeCodePosition < 6)
+      m_timeCodeStamp[m_timeCodePosition++] = remote - REMOTE_0;
+    else
+    {
+      // rotate around
+      for (int i = 0; i < 5; i++)
+        m_timeCodeStamp[i] = m_timeCodeStamp[i + 1];
+      m_timeCodeStamp[5] = remote - REMOTE_0;
+    }
+   }
+ }
+
+int CSeekHandler::GetTimeCodeAmount() const
+{
+  if (m_timeCodePosition > 0)
+  {
+    // Convert the timestamp into an integer
+    int tot = 0;
+    for (int i = 0; i < m_timeCodePosition; i++)
+      tot = tot * 10 + m_timeCodeStamp[i];
+
+    // Interpret result as HHMMSS
+    int s = tot % 100; tot /= 100;
+    int m = tot % 100; tot /= 100;
+    int h = tot % 100;
+
+    return h * 3600 + m * 60 + s;
+  }
+  return 0;
+}
+
+bool CSeekHandler::GetTimeCodeStamp(std::string *strDispTime /*= nullptr*/) const
```

Comment:
and instead of doing all formatting magic here, should we use `CDateTime` instead?

```
CDateTime time = CDateTime(static_cast<time_t>(GetTimeCodeSeconds()));
return time->GetAsLocalizedTime("HH:MM:SS");
```